Adapt `NeptuneLogger` to new `neptune-client` api

shnela opened the pull request:
## Adapt NeptuneLogger to changes in [neptune-client](https://github.com/neptune-ai/neptune-client/issues/413) api changes

* Deprecated `NeptuneLogger` was moved to separate repo
* New `NeptuneLogger` which supports new [neptune-client](https://github.com/neptune-ai/neptune-client/issues/413) is created
* Tests for new `NeptuneLogger` are implemented
* Documentation has been updated
piggyback:
* Broken `proxy` support in old `NeptuneLogger` https://github.com/neptune-ai/neptune-client/issues/413 
   * This issue was closely related to other changes in this PR


## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- For CHANGELOG separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

pep8speaks commented:
Hello @shnela! Thanks for updating this PR. 



* In the file [`pytorch_lightning/loggers/neptune.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/2a54a7d26e5cbe6aad84b336f0347270b1ac884a/pytorch_lightning/loggers/neptune.py):

> [Line 154:121](https://github.com/PyTorchLightning/pytorch-lightning/blob/2a54a7d26e5cbe6aad84b336f0347270b1ac884a/pytorch_lightning/loggers/neptune.py#L154): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (160 > 120 characters)

shnela commented:
@williamFalcon  @awaelchli @Borda @SeanNaren @carmocca @tchaton @justusschock @kaushikb11 @edenlightning 

Hi guys.
Is there anything I have to do before you can start review of this branch?
From my perspective it's done, but if I need to fix something before you can start review, or if I can help you somehow with that, don't hesitate to ask ;)

carmocca commented on .gitignore:
```diff
@@ -158,3 +158,6 @@ data
 runs
 *trace*
+
+# neptune local files
+.neptune/
```
Move this above to the section in L140 (pl tests)

shnela replied:
Done

carmocca commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -17,23 +17,34 @@
 log = logging.getLogger(__name__)
-_NEPTUNE_AVAILABLE = _module_available("neptune")
+
+if _module_available("neptune"):
+    from neptune import __version__
+
+    _NEPTUNE_AVAILABLE = __version__.startswith('0.9.') or __version__.startswith('1.')
+else:
+    _NEPTUNE_AVAILABLE = False
```
Can you do something like this instead:

```
_NEPTUNE_AVAILABLE = _module_available("neptune")
_NEPTUNE_GREATER_EQUAL_0_9 = _NEPTUNE_AVAILABLE and _compare_version("neptune", operator.ge, "0.9.0")
```

And then use these properties for imports instead of `try-except`?

Borda replied:
yeah version would be better as you never now when the new API become also Legacy :]

shnela replied:
ok

carmocca commented on README.md:
```diff
@@ -257,7 +257,7 @@ Here are some examples:
 
   # neptune
-  trainer = Trainer(logger=loggers.NeptuneLogger())
+  trainer = Trainer(logger=loggers.NeptuneLegacyLogger())
```
Is the new `NeptuneLogger` backwards compatible with the old one?

If not, we should keep the current name and instead have the new one use a different name. The current one would be deprecated and eventually removed.

Herudaio replied:
Hi @carmocca @awaelchli @Borda! 

Marcin, Product Manager from Neptune on this side. 

The context behind this change is that we introduced a new API, however, the legacy one will continue to live for some time (we plan to support it for at least a year). We are in the process of slowly sunsetting it, but it's not "deprecated" yet - we are actively maintaining it. The new API is the new default one (documentation, onboarding, etc.).

We thought about two options:
• **Adding Logger using new API as a new Logger.**
This solution has a problem that the new Logger would have a misleading name (New/Next/2/Next) for the user and the user does not know which one should they use.
• **Renaming current one as Legacy and adding new Logger under the "default" name.**
This solution on the other hand has a problem of backward compatibility as you mentioned. With the initial design for the logger we hoped to minimize this problem. However, the points about catching outdated parameters through kwargs and some methods no longer available is not a pleasant thing for a user.

As you pointed out we could improve this. I would propose the following modification:
1) If the user has not updated neptune-client we inform them about this and point to NeptuneLegacyLogger. [This is partially done, but can be improved with messaging]
2) If the user's underlying project is not migrated to the new structure we inform them about this and point to NeptuneLegacyLogger. This is temporary only, new projects are using the new structure, we are in the process of migrating existing projects. [This is partially done, but can be improved with messaging]
3) Capture old parameters through kwargs to detect "Legacy usage" and act accordingly.
4) Add methods that were exposed by base NeptuneLogger (most of the interaction was actually through `logger.experiment.method`) to keep the API of the logger the same. If the user is using those methods we act in a similar way as in 3.

Let me know what you think about that and we will act accordingly :) 

I want the transition to be as painless for the users as possible, but on the other side we did change this API based on feedback as it was often very limiting to the users and we don't want to mislead new users into using the legacy API.

awaelchli replied:
> As you pointed out we could improve this. I would propose the following modification:

Sounds good to me 😃 
@carmocca @edenlightning what do you think?

shnela replied:
Ok, adapted to your  suggestions

carmocca replied:
You could do something like this so users don't have to know about the different logger classes:

```python
class BaseNeptuneLogger: ...
class NeptuneLegacyLogger(BaseNeptuneLogger): ...
class NeptuneV2Logger(BaseNeptuneLogger): ...

def choose_logger(*args, **kwags) -> BaseNeptuneLogger:
    if on_latest_neptune_version() and latest_version_args(*args, **kwargs):
        return NeptuneV2Logger(*args, **kwargs)
    show_should_update_warning()
    return NeptuneLegacyLogger(*args, **kwargs)

NeptuneLogger = chooose_logger
```

The disadvantage of this approach is that docs will be shown under the `Neptune{Legacy,V2}Logger` classes.

shnela replied:
We've already decided that we'll take care of arguments passed to new logger.
If user will pass any deprecated arguments we'll inform him that his code is not adapted to the new one.

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -114,160 +113,214 @@ def any_lightning_module_function_or_hook(self):
+
+            from pytorch_lightning.loggers import NeptuneLogger
+
+            PARAMS = {'batch_size': 64,
+                      'lr': 0.07,
+                      'decay_factor': 0.97}
+
+            neptune_logger = NeptuneLogger(api_key='ANONYMOUS',
+                                           close_after_fit=False,
+                                           project='common/new-pytorch-lightning-integration')
```
indent using one tab

shnela replied:
ok

awaelchli reviewed (commented):
LGTM on a high level.

Borda commented on CHANGELOG.md:
```diff
@@ -119,6 +122,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed warnings and recommendations for dataloaders in `ddp_spawn` ([#6762](https://github.com/PyTorchLightning/pytorch-lightning/pull/6762/))
 
 
+- Rename `pytorch_lightning.loggers.neptune.NeptuneLogger` to `pytorch_lightning.loggers.neptune_legacy.NeptuneLegacyLogger` ([#6867](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867))
```
this shall be in deprecated section

shnela replied:
ok

Borda commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -17,23 +17,34 @@
+else:
+    _NEPTUNE_AVAILABLE = False
 
 if _NEPTUNE_AVAILABLE:
```
lets having it as both since all loggers are optional
```suggestion
if _NEPTUNE_AVAILABLE or _NEPTUNE_LEGACY_AVAILABLE:
```

shnela replied:
Legacy is version < 0.9, so you mean  `_NEPTUNE_AVAILABLE and _NEPTUNE_GREATER_EQUAL_0_9`?

Borda commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -114,160 +113,214 @@ def any_lightning_module_function_or_hook(self):
-        prefix: str = '',
-        **kwargs
-    ):
+            self,
+            api_key: Optional[str] = None,
+            project: Optional[str] = None,
+            close_after_fit: Optional[bool] = True,
+            name: Optional[str] = None,
+            run: Optional[str] = None,
+            prefix: str = '',
+            **neptune_run_kwargs):
```
ok this will be compatibility/breaking change as users would import the sea class as before but no it has very different args

Borda commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -114,160 +113,214 @@ def any_lightning_module_function_or_hook(self):
+            name: Optional[str] = None,
+            run: Optional[str] = None,
+            prefix: str = '',
+            **neptune_run_kwargs):
```
as this is breaking, you should add check if `neptune_run_kwargs` contains any of past API args and raise configuration error... The scenario is that a user upgrade Pl but still has code ith legacy Logger but the new Clas name match the old one so he wouldn't notice anything and as he is passing old args they won't be recognized and fall into `neptune_run_kwargs` basket and in fact, the logger will run with defaults

shnela replied:
yes, we're checking if user uses deprecated values and warn user that he has deprecated implementation.

Borda reviewed (changes requested):
Pls, consider rather use a new name for this new logger than rename the old Legacy...

shnela commented:
@awaelchli  @Borda 

Ok, everything updated and ready to merge.

awaelchli commented:
Thanks a lot! We will follow up here as soon as 1.3 is released (May 5ish). We currently have feature freeze for stability testing and will not be able to merge new features and refactors.

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -114,160 +140,229 @@ def any_lightning_module_function_or_hook(self):
         params = self._flatten_dict(params)
         for key, val in params.items():
-            self.experiment.set_property(f'param__{key}', val)
+            self.run[f'param__{key}'] = val
```
@Herudaio do we want naming scheme `param__{key}` in the new logger? Why not `params/{key}`?

PiotrJander replied:
Also, @shnela cover with tests

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -114,160 +140,229 @@ def any_lightning_module_function_or_hook(self):
             # `step` is ignored because Neptune expects strictly increasing step values which
             # Lighting does not always guarantee.
-            self.log_metric(key, val)
+            self.experiment[key].log(val)
```
`self.run[key].log(val)`

PiotrJander replied:
@Herudaio do we not want to log metric in a namespace (e.g. `logs`) instead of logging them at the root of the run?

PiotrJander replied:
Also, @shnela cover with tests

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -276,118 +371,50 @@ def save_dir(self) -> Optional[str]:
-        """
-        self.experiment.log_text(log_name, text, step=step)
+    def log_metric(self, *args, **kwargs):
+        self._raise_deprecated_api_usage("log_metric", "logger.run[constants.LOG_ATTRIBUTE_SPACE]['key'].log(42)")
```
Do we want the user to see `constants.LOG_ATTRIBUTE_SPACE`? Or just `logger.run['log/key'].log(42)`

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -276,118 +371,50 @@ def save_dir(self) -> Optional[str]:
+    @rank_zero_only
+    def append_tags(self, *args, **kwargs):
+        self._raise_deprecated_api_usage("append_tags",
+                                         "logger.run[constants.SYSTEM_TAGS_ATTRIBUTE_PATH].add(tag/tags)")
```
`logger.run['sys/tags'].add(['foo', 'bar'])`

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -226,6 +230,7 @@ def __init__(
             name: Optional[str] = None,
             run: Optional[str] = None,
             prefix: str = '',
+            base_namespace: str = None,
```
Set the default value to an empty string `''` as opposed to `None`, this will simplify the rest of the code.

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -337,9 +343,12 @@ def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
-        params = self._flatten_dict(params)
-        for key, val in params.items():
-            self.run[f'param__{key}'] = val
+
+        parameters_key = self.PARAMETERS_KEY
+        if self._base_namespace:
+            parameters_key = f'{self._base_namespace}/{parameters_key}'
+
+        self.run[parameters_key] = params
```
Just do `self.run[self._base_namespace][parameters_key] = params` assuming that `self._base_namespace` is an empty string by default

shnela replied:
ok

shnela replied:
One the other hand it will make tests significantly more complicated and ugly, are you sure that we don't want to keep this internal if?

PiotrJander commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -353,10 +362,14 @@ def log_metrics(self, metrics: Dict[str, Union[torch.Tensor, float]], step: Opti
             # `step` is ignored because Neptune expects strictly increasing step values which
             # Lighting does not always guarantee.
-            self.experiment[key].log(val)
+            self.experiment[f'{metrics_key}/{key}'].log(val)
```
`self.run[self._base_namespace][metrics_key][key].log(val)`

shnela replied:
ok

Herudaio commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -199,12 +230,23 @@ def __init__(
+        if used_legacy_kwargs:
+            raise ValueError(
+                f"Following kwargs used by you are deprecated: {used_legacy_kwargs}."
+                f" You should use arguments accepted by either `NeptuneLogger.init` or `neptune.init`."
```
@shnela let's change this to:
Following kwargs used by you are deprecated: {used_legacy_kwargs}.
If you are looking for the Neptune logger using legacy Python API it has been renamed to NeptuneLegacyLogger. The NeptuneLogger was re-written to use the neptune.new Python API (learn more: https://neptune.ai/blog/neptune-new).
You should use arguments accepted by either `NeptuneLogger.init` or `neptune.init`

shnela replied:
ok

Herudaio commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -342,3 +397,35 @@ def version(self) -> str:
         return self.run['sys/id'].fetch()
+
+    def _raise_deprecated_api_usage(self, f_name, sample_code):
+        raise ValueError(f"Function you've used  is deprecated."
```
Change to:
The function you've used is deprecated.
If you are looking for the Neptune logger using legacy Python API it has been renamed to NeptuneLegacyLogger. The NeptuneLogger was re-written to use the neptune.new Python API (learn more: https://neptune.ai/blog/neptune-new).
Instead of `logger.{f_name}` you can use:
\t{sample_code}

shnela replied:
ok

shnela commented:
Hi @awaelchli 

I see that you've succeed with publishing `neptune-pytorch 1.3.0`, congrats!

I wonder how does your policy regarding new features look like. Is it possible to publish `neptune` integration in release `1.3.1`, when do you plan to publish such release?

Best
Jakub

awaelchli commented on tests/loggers/test_neptune.py:
```diff
@@ -119,8 +67,115 @@ def _run_training(logger):
+        'notebook_path',
+    ]
+    for legacy_kwarg in legacy_neptune_init_kwargs:
+        _assert_legacy(
+            NeptuneLogger,
+            **{legacy_kwarg: None}
+        )
```
I think you could do something like this directly:
```suggestion
        with pytest.raises(ValueError, match="your message"):
            NeptuneLogger(...)
```

awaelchli commented on tests/loggers/test_neptune.py:
```diff
@@ -119,8 +67,115 @@ def _run_training(logger):
+
+
+def _assert_legacy(callback, *args, **kwargs):
+    try:
+        callback(*args, **kwargs)
+    except ValueError:
+        pass
+    else:
+        raise AssertionError("Should throw `ValueError`")
```
I think the pytest way is to do this:

```python
with pytest.raises(ValueError, match="your message"):
            callback(...)
```

shnela replied:
Yes, fixed

awaelchli reviewed (commented):
overall LGTM

awaelchli commented:
Thanks Jakub

> I wonder how does your policy regarding new features look like. Is it possible to publish neptune integration in release 1.3.1, when do you plan to publish such release?

We reserve 1.3.x for bugfixes exclusively, the don't have new features or refactors. 
I contacted @edenlightning to figure out what we can do. Will get back to you

edenlightning commented:
Since this isn't a bugfix, it can be included in the next minor release (1.4), expected to be released in 10-12 weeks.

kamil-kaczmarek commented:
hey @edenlightning ,

Kamil from neptune.ai, nice to meet you here.

On the Neptune side we now work on the improved integration. Do you have any estimate when you are going to release 1.4? I'm trying to figure out deadlines for our work?

Thanks.

kamil-kaczmarek commented:
hey @edenlightning ,

This is friendly reminder about my previous questions:

* Do you have any estimate when you are going to release 1.4?
* I'm trying to figure out deadlines for our work?

shnela commented:
@williamFalcon @awaelchli @Borda @SeanNaren @carmocca @tchaton @justusschock @kaushikb11 @edenlightning

Hi again :)

Unfortunately [we wasn't able to merge this changes and release as part of PTL1.4](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#issuecomment-834265844).

Since then we've cleaned this PR recently and updated our new `NeptuneLogger` a little bit.

For us it's extremely important to release these changes in version 1.5.0, so they will be available in September.
So we'd be grateful if you've reviewed this branch and give us feedback about what we should improve to make it happen.

From our perspective the only thing we'll have to is updating docstrings in new `NeptuneClient`.

So, is there something you want us to improve? :)

awaelchli commented:
Great! Love it. When it's ready for review feel free to ping us @shnela

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -11,29 +12,80 @@
+]
 
 
 class NeptuneLogger(LightningLoggerBase):
```
Mind adding as code-owner for this file: @shnela

shnela replied:
Yes, I'll definitely add our team as code owners, thanks for suggestion!

tchaton replied:
Yes, so you get notified to issues and maybe we can find a way to triage issues directly to your team too.

tchaton replied:
We will create a netpune label, so it is easier for your team to navigate issues.

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,235 +98,437 @@ class NeptuneLogger(LightningLoggerBase):
 
-    .. code-block:: python
+    Check `Neptune docs <https://docs.neptune.ai/user-guides/logging-and-managing-runs-results/logging-runs-data>`_
+    for more info about how to log various types metadata (scores, files, images, interactive visuals, CSVs, etc.).
```
Would be neat for Netpune to natively add support for this as `log_image, log_csv, log_...`. Maybe another PR.

shnela replied:
Ok, since you don't have clear interface in [Standardize API for logging proposal](https://github.com/PyTorchLightning/pytorch-lightning/issues/6720), I'll support only method which were available in previous version of `NeptuneClient`.

Once you decide what set of `log_*` functions you want to keep, we'll implement it :)

tchaton replied:
Sounds great ! Thanks.

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,235 +98,437 @@ class NeptuneLogger(LightningLoggerBase):
+            name: Optional[str],
+            run: Optional["Run"],
+            neptune_run_kwargs: dict):
+
```
Add a comment about this function and previous client.

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,235 +98,437 @@ class NeptuneLogger(LightningLoggerBase):
+        return set(cls._dict_paths(uploaded_models_dict))
+
+    @classmethod
+    def _dict_paths(cls, d: dict, path_in_build: str = None):
```
typing ?

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -284,9 +538,7 @@ def name(self) -> str:
-        if self.offline_mode:
-            return "offline-name"
-        return self.experiment.name
+        return "NeptuneLogger"
```
Why `NeptuneLogger` there ? This is a BC.

awaelchli replied:
I think we should return the actual experiment name here by querying the client

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -284,9 +538,7 @@ def name(self) -> str:
         Returns:
             The name of the experiment if not in offline mode else "offline-name".
         """
-        if self.offline_mode:
```
Won't the `offline_mode` be supported anymore ?

shnela replied:
We still suuport `offline_mode`, but we don't have to check it since function always returns "NeptuneLogger".

The function comment is not up to date, but our product team is currently working on docstrings ;)

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -296,113 +548,42 @@ def version(self) -> str:
-            self.experiment.log_metric(metric_name, metric_value)
-        else:
-            self.experiment.log_metric(metric_name, x=step, y=metric_value)
+    def log_metric(self, *args, **kwargs):
```
Why is this depreciated ?

tchaton replied:
This is not a path we want to take for now.

All the loggers are exposing those functions and Netpune Logger shouldn't be different as it will involved learning a new API for Lightning users.

Please, add support back for all the `log_{}` and maybe add a warning that users can use log directly if needed.

shnela replied:
I don't see any definitions of `log_metric`, `log_text`, `log_image` etc... in any other loggers.

Those methods were available in previous version of logger and we've kept them to inform user that this is deprecated usecase only.

shnela replied:
It's related to [this comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r699224411) I guess.

tchaton replied:
Yes, your integration is definitely in advance compared to other loggers.
This syntax is related to this issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/6720

shnela replied:
Ok, if you have official plans for that I see no problems with keeping support for those methods.
I'll try to implement them and add some tests.

Herudaio replied:
@tchaton 

> Yes, your integration is definitely in advance compared to other loggers.
> But we are in the process to add this `log_{}` for all loggers
> The issue is there: #6720
> Let's make sure that using Netpune `log_{}` works and don't raise an error.

To be sure when you said "works and don't raise an error." do you mean:
a) Remove the implementation (prepared for adding it in the future, existing users will get an error)
b) Keep the method, don't raise an error, print warning informing that if they are looking for the previous implementation they can find here (in essence what is now, but without the first sentence).
c) Keep the method, with empty implementation
d) Something else?

Once the API is there we will be definitely implementing those on our side, but as far as I understand the API is not yet finalized?

Herudaio replied:
@tchaton ping on the question here ☝🏻 :)

awaelchli replied:
is it possible there was a confusion here between `log_metric` (neptune specific) and `log_metrics` (interface of the base class)?

shnela replied:
> is it possible there was a confusion here between log_metric (neptune specific) and log_metrics (interface of the base class)?

That was my initial guess, but after @tchaton send us https://github.com/PyTorchLightning/pytorch-lightning/issues/6720  issue I guess that some of the methods like: `log_metric`, `log_text` or `log_image` may be part of the logging API, so we wanted to avoid situation when we:
* first implemented those methods in our old `NeptuneLogger`
* then make them deprecated in our new `NeptuneLogger`
* and then revoke them again as they will be part of new Logger API.

It'd be nice if we implemented them now, but we can't since we don't know list of exptected arguments.

So maybe the best solution is just keeping raise of exception when methods are used, but instead of information that they are deprecated, we could add some more detailed information that they were part of old `NeptuneLogger`, which currently is available in separate repo (according to your request), but we could also mention that those methods **may** be part of some future API.

Eventually we can just drop those functions, so they are not part of new `NeptuneLogger` at all...
This deprecation info is not very useful in fact since someone who imports new `NeptuneLogger` and runs it against old code, will be warned after passing wrong arguments to `__Init__` function.

What do you think?

tchaton replied:
Yes, I believe `log_metrics` is used across Lightning. This should be the API proposed by the NetpuneLogger.

@Herudaio

What I meant is `log_artifact`, `log_image`, etc... were supported in the previous implementation and should still be supported with the new one.

Otherwise, it would be a big BC change for the users.

Best,
T.C

Herudaio replied:
> I don't believe a warning would even be required there as they shouldn't be a depreciation.
> The users are relying on the Lightning API of a logger, not the Neptune API.
> 
> Best,
> T.C

Hmm, the initial addition of `log_artifact`, `log_image` on the logger side was the aim of 1:1 parity with Neptune API. As @shnela mentioned at that time there was no Lightning API for that. 

@tchaton could you point us what API (parameters, return type) for those `log_{}` methods should look like? We only found one closed pull request with API proposition for `log_images`, but maybe we are not aware of something. Could you provide a link where we can find what the method signatures should be?

awaelchli replied:
I think all we want (if I understand @tchaton correctly) is that you handle either one of these cases:

a) your intention is to remove log_x -> deprecate it in 1.5, remove in version 1.7 but until then provide the functionality if possible
b) your intention is not to remove it -> then keep api the same, no action required

The priority here is to make it as backward compatible as possible. The whole discussion about what log_x might look like in the future for Lightning is out of scope in my opinion can be discussed separately elsewhere.

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,235 +98,437 @@ class NeptuneLogger(LightningLoggerBase):
+    @staticmethod
+    def _get_full_model_name(model_path: str, checkpoint_callback: "ReferenceType[ModelCheckpoint]") -> str:
+        """Returns model name which is string `modle_path` appended to `checkpoint_callback.dirpath`.
+        The problem here is that model name may contain slashes('/')."""
```
Any idea how to resolve this ?

shnela replied:
It's resolved, the comment is explanation why such logic in `_get_full_model_name` was required at all, but it's misleading in fact.
I'll just remove this comment.

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,235 +98,437 @@ class NeptuneLogger(LightningLoggerBase):
+        ]
+        if used_legacy_neptune_kwargs:
+            raise ValueError(
+                # TODO: product - review text
```
Any updates there ?

shnela replied:
Yes, @kamil-kaczmarek from our product team is currently working on updating docstrings and reviewing messages displayed to user.
The commit will be published today.

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,235 +98,437 @@ class NeptuneLogger(LightningLoggerBase):
+            )
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
```
Does this ensure reloading from previous checkpoint ?

tchaton replied:
I am not familiar enough, but you have have a test in mind, it would be good :)

shnela commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -1,4 +1,5 @@
-# Copyright The PyTorch Lightning team.
+#
```
Right, sorry I've copy pasted this from our repo...

shnela commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -11,29 +12,80 @@
-if _NEPTUNE_AVAILABLE:
-    import neptune
-    from neptune.experiments import Experiment
+_NEPTUNE_GREATER_EQUAL_0_9 = _NEPTUNE_AVAILABLE and _compare_version("neptune", operator.ge, "0.9.0")
```
ok

shnela commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -11,29 +12,80 @@
+INTEGRATION_VERSION_KEY = "source_code/integrations/pytorch-lightning"
+
+# kwargs used in NeptuneLogger for legacy client (current NeptuneLegacyLogger)
+LEGACY_NEPTUNE_INIT_KWARGS = [
```
Sorry I don't get this, this is list of arguments which were available in previous `NeptuneLogger` and are not available in current `NeptuneLogger`.
Where should I store them?

I can make it a private list of `NeptuneLogger` if you want.

tchaton replied:
I meant, I think you could do:

```py
from neptune.pytorch_lightning import _ INTEGRATION_VERSION_KEY, _LEGACY_NEPTUNE_INIT_KWARGS, ...
```

shnela replied:
I see.
But I don't really think that it's good idea to keep those variables in separate repository.
We'd have to update our `neptune-client`, release it and then add proper version as minimal supported dependency to `pytorch-lightning` library.
It could make some sense if `LEGACY_NEPTUNE_INIT_KWARGS` would be used in several places, then `neptune.pytorch_lightning` would be the source of true.
But this list keeps difference between args accepted by previous version of `NeptuneLogger` and current version of `NeptuneLogger` I don't see any reason why should we keep it in other repo.

`INTEGRATION_VERSION_KEY` is something we always keep as global variable in our integrations code.

I agree that we could make those variables "module private" by prefixing them with `"_"`, or add `__all__` statement which will make only `NeptuneLogger` importable.

shnela replied:
@tchaton what do you think?

shnela commented:
@awaelchli from our perspective almost everything is done.
@tchaton asked several PR questions and we've fixed almost all of them. There are only three unresolved issues we're discussing:
* [Checkpoint reloading](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r699231981) - we need more clarification
* [Supporting log_* methods](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r699227852) - we're quite confused about that one. Any other logger implements it.

shnela commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,247 +93,469 @@ class NeptuneLogger(LightningLoggerBase):
+        return set(cls._dict_paths(uploaded_models_dict))
+
+    @classmethod
+    def _dict_paths(cls, d: dict, path_in_build: str = None):
```
done

awaelchli commented on .github/CODEOWNERS:
```diff
@@ -25,6 +25,7 @@
 /pytorch_lightning/core                 @tchaton @SeanNaren @borda @carmocca @justusschock @kaushikb11
 /pytorch_lightning/distributed          @williamfalcon @tchaton @awaelchli @kaushikb11
 /pytorch_lightning/loggers              @tchaton @awaelchli @borda
+/pytorch_lightning/loggers/neptune.py   @shnela @HubertJaworski @pkasprzyk @pitercl @Raalsky @aniezurawski
```
great!

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -15,25 +15,72 @@
+]
+
+# kwargs used in legacy NeptuneLogger from neptune-pytorch-lightning package
+LEGACY_NEPTUNE_LOGGER_KWARGS = [
```
if what @tchaton suggests isn't possible then at least we should make it protected :) 
```suggestion
_LEGACY_NEPTUNE_LOGGER_KWARGS = [
```

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,247 +93,469 @@ class NeptuneLogger(LightningLoggerBase):
+                metadata = ...
+                self.logger.experiment["your/metadata/structure"].log(metadata)
+
+    Check `Logging metadata docs <https://docs.neptune.ai/you-should-know/logging-metadata>`_
```
not a big deal, just nitpick: I suggest to put the extensive user guide into the neptune.rst docs and keep all the docs here concise and focused around the api. but up to you

shnela replied:
@kamil-kaczmarek @Herudaio

kamil-kaczmarek replied:
@awaelchli happy to move more extensive user guide to the neptune.rst. Would you mind to recommend a location of it in the PTL docs? Seems that `tutorials` is good candidate.

awaelchli replied:
we can do in a follow up if necessary, shouldn't block this pr

kamil-kaczmarek replied:
@awaelchli agree, let not block this pr. I will update it after the release. Maybe you can include it in the weekly patch release.

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,247 +93,469 @@ class NeptuneLogger(LightningLoggerBase):
-        Log metrics (numeric values) in Neptune experiments.
+        Hyperparams will be logged under the "<prefix>/hyperparams" namespace.
+
+        **Note**
+
+        You can also log parameters by directly using the logger instance:
+        ``neptune_logger.experiment["model/hyper-parameters"] = params_dict``.
+
+        In this way you can keep hierarchical structure of the parameters.
```
```suggestion
        Note:

            You can also log parameters by directly using the logger instance:
            ``neptune_logger.experiment["model/hyper-parameters"] = params_dict``.

            In this way you can keep hierarchical structure of the parameters.
```
for google-style sphinx formatting we use the predefined sections

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,247 +93,469 @@ class NeptuneLogger(LightningLoggerBase):
+
+        self.run[parameters_key] = params
+
+    @rank_zero_only
+    def log_metrics(self, metrics: Dict[str, Union[torch.Tensor, float]], step: Optional[int] = None) -> None:
+        if rank_zero_only.rank != 0:
+            raise ValueError("run tried to log from global_rank != 0")
```
there is a decorator here preventing this, at least in lightning.
did you add the error handling here expecting users to call this method manually?

shnela replied:
Yes I think it's redundant check.

shnela replied:
I've just made assertion here, but checkstyle forced me to change it to raising an Exception.

It's here because when something goes wrong and we'll try to log to the same logger from several threads, `neptune-client` may crash due to non synchronized timestamps sent to the series.
It's very unlikely that this happens because of mentioned `@rank_zero_only` but I've used this defensive strategy since it's much more easier to track such error from exception rather than mentioned problem with logging.

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -296,113 +565,46 @@ def version(self) -> str:
-        if self.offline_mode:
-            return "offline-id-1234"
-        return self.experiment.id
+        return self.run._short_id  # skipcq: PYL-W0212
```
```suggestion
        return self.run._short_id
```
what's this?

shnela replied:
Ok, I see that we have problem here...

`Neptune` has project with attached experiments.
Let's assume that we have and project with name `Domestic Market` with abbreviation `DOM`.

Then experiments created for this project will have auto generated short ids like follows: `DOM-1`, `DOM-2`, `DOM-3`...

So I guess that `logger.name` should return those values accordingly instead of "NeptuneLogger".
And `logger.version` should return the same values or integers like: 1, 2, 3... Which one do you prefer?

awaelchli replied:
no I meant the skipcq comment there :)

shnela replied:
Oh, it's here because your sytylecheck complained on using "private" attribute of `run`.

Unfortunately we have to do it his way, because `_short_id` is always available and doesn't require additional API CALL to obtain id.

awaelchli commented on tests/loggers/test_all.py:
```diff
@@ -233,7 +233,6 @@ def name(self):
         CometLogger,
         CSVLogger,
         MLFlowLogger,
-        NeptuneLogger,
```
so the neptune logger is not pickleble? if that's the case, please make it clear in the docs that any of the ddp variants that use spawn do not support running neptune logger. 

let me know if my comment is unclear, I can add more explanation if necessary

shnela replied:
I've removed it from this general test, because most of `NeptuneLogger` should mock `neptune-client` otherwise they don't work.
I didn't want to adapt general test case for particular logger.

But yes, there's no tests for pickling `NeptuneLogger`...
But our logger should be pickleable though. We're removing non pickleable `_run_isntantce` object in `__getstate__` method.

Testing if we are able to pickle and unpickle `NeptuneLogger` instance in will be enough, or we must create more sophisticated tests like `test_wandb_pickle`?

awaelchli replied:
a simple test like this would do perfectly fine (pseudo code):

```python
@RunIf(NeptuneAvailable)
def test_neptune_pickle():
logger = NeptuneLogger()
_ = logger.experimenet
pickle.dump(logger)  # should not error :) 
```

not a strong requirement from my side, but would be nice to have.

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -296,113 +565,46 @@ def version(self) -> str:
-        """
-        self.experiment.set_property(key, value)
+    def set_property(self, *args, **kwargs):
+        self._raise_deprecated_api_usage("log_artifact",
+                                         f"logger.run['{self._prefix}/{self.PARAMETERS_KEY}/key'].log(value)")
```
formatting broken here?

shnela replied:
Why?

awaelchli replied:
I mean it's not according to our formatting rules, there is arbitrary empty space here. 
Can you run black on the file? See also the pre-commit hook is failing :)

shnela replied:
Yes, I've just fixed all issues locally and commited changes. Thanks for info.

awaelchli replied:
ah, you are first time contributor so you didn't see our failed checks sorry. 
we had to start the CI manually (always when new contributor sends a PR).

tchaton commented on docs/source/common/loggers.rst:
```diff
@@ -129,12 +133,16 @@ The :class:`~pytorch_lightning.loggers.NeptuneLogger` is available anywhere exce
-            self.logger.experiment.add_image("generated_images", some_img, 0)
+            # generic recipe for logging custom metadata
+            metadata = ...
+            self.logger.experiment["your/metadata/structure"].log(metadata)
```
Mind adding a comment this API won't work with a collection of loggers.

shnela replied:
@Herudaio it's something for you.

kamil-kaczmarek replied:
@tchaton @shnela @Herudaio comment added

shnela commented:
@awaelchli @tchaton 
Status for today.

What is not resolved:
* Some documentation issues - @kamil-kaczmarek @Herudaio will take care of it tomorrow morning
  * [suggestion of putting docs to neptune.rst](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r700026944)
  * [info about not working collection of loggers](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r700084720)
* (from yesterday) [Supporting log_* methods](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r700142229) - we've suggested some solutions and we're waiting for detailed answer from you.
* (pickling test)[https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r700036278] - I'm waiting for information regarding pickle testing for `NeptuneLogger` but if I get no response for tomorrow I'll start implementing something similar to `test_wandb_pickle`

I hope you guys remember that it's really important for us to release these changes with pytorch-lightning 1.5.0?
Do you think we can have any trouble with that?

Best,
Jakub

awaelchli commented:
> I hope you guys remember that it's really important for us to release these changes with pytorch-lightning 1.5.0?
> Do you think we can have any trouble with that?

I don't see anything standing in the way of that.

tchaton commented:
Hey @shnela,

Let's remove the `_raise_deprecated_api_usage` and make sure the functionality is supported internally so it doesn't introduce a BC.

Best,
T.C

shnela commented:
Hi @awaelchli ,

it looks like we're almost done with this PR.
All we need is information [from you](https://github.com/PyTorchLightning/pytorch-lightning/pull/6867#discussion_r701235084) about neptune PTL docs.

Best,
Jakub

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,355 +94,557 @@ class NeptuneLogger(LightningLoggerBase):
+                run_instance = neptune.init(
+                    project=project,
+                    api_token=api_key,
+                    name=name,
```
You could add some metadata like `source = lightning`, so you would know how many users are using NetptuneLogger from PyTorch Lightning.

Herudaio replied:
@tchaton Thanks for the suggestion! Actually we are already doing that with this line:
`run_instance[_INTEGRATION_VERSION_KEY] = __version__`

shnela replied:
We're already doing it by saving information about version in `_INTEGRATION_VERSION_KEY`

tchaton commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,355 +94,557 @@ class NeptuneLogger(LightningLoggerBase):
-        """Log metrics (numeric values) in Neptune experiments.
+        if not raise_exception:
+            warnings.warn(
+                "The function you've used is deprecated and will be shut down in pytorch-lightning 1.7.0.\n"
```
```suggestion
                "The function you've used is deprecated in v1.5.0 and will be removed in v1.7.0. "
```

tchaton reviewed (approved):
LGTM !

awaelchli commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,355 +94,557 @@ class NeptuneLogger(LightningLoggerBase):
-            self.experiment.log_image(log_name, x=step, y=image)
+        key = f"{self._prefix}/{log_name}"
+        self._signal_deprecated_api_usage("log_image", f"logger.run['{key}'].log(File('path_to_image'))")
+        if isinstance(image, str):
+            # if `img` is path to file, convert it to file object
+            img = NeptuneFile(image)
+        self.run[key].log(img, step=step)
```
```suggestion
        if isinstance(image, str):
            # if `img` is path to file, convert it to file object
            image = NeptuneFile(image)
        self.run[key].log(image, step=step)
```
local variable 'img' referenced before assignment

tchaton commented:
Hey @shnela,

There is one failing test.

Best,
T.C

shnela commented:
@awaelchli @tchaton is there anything I can do for you guys?
Or maybe I can just wait for PR to be merged :P

shnela commented:
@tchaton fixed.

shnela commented:
@tchaton and again fixed `Docs check / test-docs (pull_request) `.

awaelchli commented:
see [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/8584#discussion_r703566492), maybe you need to run this command.

shnela commented:
@awaelchli @tchaton should be good now.

awaelchli commented:
Thanks. Can you merge master so the branch is in sync? For the next PR it would be good if you could allow us maintainers to make changes to the branch, then we can better help out with these pre-commit checks and so on.

shnela commented:
@awaelchli 
> Thanks. Can you merge master so the branch is in sync?

I've just rebased branch and pushed changes. I think we should squash commits from this branch before merging (github can automatically do that), since the commit history is quite messy here and there's no merge commits.

> For the next PR it would be good if you could allow us maintainers to make changes to the branch, then we can better help out with these pre-commit checks and so on.

Yes! We definitely should and will do that!

awaelchli commented:
yes every pr gets merged with squash. no history will be carried over from this branch to master.

ethanwharris commented on pytorch_lightning/loggers/neptune.py:
```diff
@@ -46,355 +94,556 @@ class NeptuneLogger(LightningLoggerBase):
+        ]
+        if used_legacy_kwargs:
+            raise ValueError(
+                f"Following kwargs are deprecated: {used_legacy_kwargs}.\n"
+                "If you are looking for the Neptune logger using legacy Python API,"
+                " it's still available as part of neptune-contrib package:\n"
+                "  - https://docs-legacy.neptune.ai/integrations/pytorch_lightning.html\n"
+                "The NeptuneLogger was re-written to use the neptune.new Python API\n"
+                "  - https://neptune.ai/blog/neptune-new\n"
+                "  - https://docs.neptune.ai/integrations-and-supported-tools/model-training/pytorch-lightning\n"
+                "You should use arguments accepted by either NeptuneLogger.init() or neptune.init()"
```
Could de-duplicate the code here a bit, ideally this message would only need to be defined once

ethanwharris reviewed (commented):
Looks good, minor comment :smiley:

shnela commented:
@awaelchli @tchaton so can we merge now? :p

awaelchli commented on .github/CODEOWNERS:
```diff
@@ -25,6 +25,7 @@
 /pytorch_lightning/distributed          @williamfalcon @tchaton @awaelchli @kaushikb11
 /pytorch_lightning/loggers              @tchaton @awaelchli @borda
 /pytorch_lightning/loggers/wandb.py     @borisdayma
+/pytorch_lightning/loggers/neptune.py   @shnela @HubertJaworski @pkasprzyk @pitercl @Raalsky @aniezurawski @kamil-kaczmarek
```
```suggestion
```
@carmocca  suggested to do this in a separate PR, then we can merge this PR without waiting for admin to approve. (codeowners update requires admin manual merge)

awaelchli replied:
we don't have edit access to your branch, you should see a "commit suggestion" button here so please click it :) 
I will enable auto-merge

shnela replied:
done

shnela replied:
Separate PR for updating code owners: https://github.com/PyTorchLightning/pytorch-lightning/pull/9439

shnela replied:
Sorry, I didn't noteiced you've just created one.

tchaton commented:
Hey @shnela,

Great work there ! Really excited about this :)

Best,
T.C

shnela commented:
@tchaton 
> Great work there ! Really excited about this :)

We to! Thank you for review and your help!

Best,
Jakub

kamil-kaczmarek commented:
Hi @aarzchan ,

In the new logger you can do it by parameterizing your Logger, like this (`mode="offline"`):

```
neptune_logger = NeptuneLogger(
    api_key="ANONYMOUS",  # replace with your own
    project="WORKSPACE/PROJECT",
    mode="offline",
)
```

There are more [modes](https://docs.neptune.ai/api-reference/neptune#init) available.
Overall you pass any neptune.init() argument to the `NeptuneLogger` and in this way further define its' behavior.

raman-rajarathinam commented:
Hi @shnela @kamil-kaczmarek ,
Is it possible to use old neptune API when using master? If yes, are there any examples?

kamil-kaczmarek commented:
Hi @raman-rajarathinam,

We are preparing an update to the neptune [legacy-docs](https://docs-legacy.neptune.ai/), where this use case will be supported and described.

We are planning to release an update next week.

i008 commented:
Hi @kamil-kaczmarek after upgrading to pl==1.5.0 Neptune stopped working for me:


When i start it it print the experiment multiple times and keeps printing the messege as below.

https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
https://app.neptune.ai/i008/bovine/e/BOV-575
Global seed set to 1245
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:48.990Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:48.992Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:49.359Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:49.361Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:49.722Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:49.726Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:50.099Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:50.105Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:50.477Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:50.481Z
Error occurred during asynchronous operation processing: Timestamp must be non-decreasing for series attribute: monitoring/stdout. Invalid point: 2021-11-05T10:06:50.840Z

kamil-kaczmarek commented:
Hey @i008,

Did the same code worked well before upgrade to `1.5` ?

To confirm: is it training on many machine or with muliple processes?

i008 commented:
One local gpu. The (almost) same code worked. I had to adjust the init of the logger.

kamil-kaczmarek commented:
@i008,

* Do you use `CUSTOM_RUN_ID` or pass the run object around in your code base?
* Can you share what you modified in the init?

Problem is that async processes write to the same place (`stdout` in your case) in Neptune and this causes these errors msgs. In practice it does not break your run, these points are simply skipped.

There are few strategies to mitigate that. Happy to chat online if it's ok for you.

i008 commented:
Hey @kamil-kaczmarek 
Whats the best way to get in touch online? Maybe you can send me a hangout invite @  kubacieslik@gmail.com Im avaialble next ~3 hours if that works for you.

i008 commented:
I did some further debugging and it seems that this i related to the CLI interface. The problem only occurs while using neptune like that:

```
    neptune_logger = NeptuneLogger(
        api_key=NEPTUNE_KEY,
        project='projec',
        source_files=["**/*.py", "**/*.ipynb"],

    )
    
    train_defaults = {
        'logger': neptune_logger,
        'gpus': 1,
        'precision': 16

    }
    cli = CLI(
        LitAutoEncoder,
        Data,
        trainer_defaults=train_defaults,
    )

```

If i do it the regular way ( trainer = , trainer.fit())  things work as expected.

kamil-kaczmarek commented:
Great, thanks for the info. Let's dicsuss more online

tshu-w commented:
@kamil-kaczmarek Hi, should we add a init args `version` to the NeptuneLogger so that the user can customize the custom_run_id and align it with other Loggers in PytorchLightning like [TensorBoardLogger](https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.loggers.TensorBoardLogger.html#pytorch_lightning.loggers.TensorBoardLogger) and [WandbLogger](https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.loggers.WandbLogger.html#pytorch_lightning.loggers.WandbLogger)?

carmocca commented:
@shnela @kamil-kaczmarek The deprecated code can now be removed (`_signal_deprecated_api_usage`)
